lib/ExtraCorres.thy

```diff
@@ -293,82 +295,98 @@ lemma corres_whileLoop:
-                  and R="{((r', s'), r, s). C' r s \<and> (r', s') \<in> fst (B' r s)
-                                            \<and> whileLoop_terminates C' B' r s}"
-                  in not_snd_whileLoop)
+          (fastforce intro: body_corres body_inv rrel dest: nf cond))
+  apply (rule_tac I="\<lambda>rv' s'. \<exists>rv s. (s, s') \<in> srel \<and> rrel rv rv' \<and> P rv s \<and> P' rv' s'"
+              and R="{((r', s'), r, s). C' r s \<and> (r', s') \<in> fst (B' r s)
+                                        \<and> whileLoop_terminates C' B' r s}"
+               in not_snd_whileLoop)
     apply (fastforce intro: rrel)
-   apply (rename_tac conc_r s)
+   apply (rename_tac s s' conc_r new_s)
    apply (clarsimp simp: validNF_def)
    apply (rule conjI)
     apply (intro hoare_vcg_conj_lift_pre_fix; (solves wpsimp)?)
-      apply (prop_tac "\<exists>abs_r. rrel abs_r conc_r")
-       apply (fastforce simp: rrel2)
-      apply clarsimp
+      apply (rule_tac Q="\<lambda>s'. \<exists>rv s. (s, s') \<in> srel \<and> rrel rv conc_r
+                                     \<and> P rv s \<and> (P' conc_r s' \<and> C' conc_r s') \<and> s' = new_s"
+                   in hoare_weaken_pre[rotated])
+       apply clarsimp
+      apply (rule hoare_ex_pre)
+      apply (rename_tac abs_r)
       apply (rule hoare_weaken_pre)
-       apply (fastforce intro!: wp_from_corres_u body_inv body_corres)
+       apply (rule_tac G="rrel abs_r conc_r" in hoare_grab_asm)
+       apply (wpsimp wp: wp_from_corres_u[OF body_corres] body_inv)
+       apply (fastforce dest: nf)
       apply (fastforce dest: cond)
      apply (fastforce simp: valid_def)
     apply wpsimp
     apply (rule whileLoop_terminates_cross[OF body_corres];
-           (fastforce dest: cond intro: body_inv abs_termination))
-   apply (prop_tac "\<exists>abs_r. rrel abs_r conc_r")
-    apply (fastforce simp: rrel2)
-   apply clarsimp
-   apply (rule_tac P="\<lambda>s'. \<exists>s. (s, s') \<in> srel \<and> (P and C abs_r) s \<and> P' s' \<and> C' conc_r s'"
-                   in no_fail_pre)
-    apply (insert cond body_corres)
-    apply (fastforce intro: corres_u_nofail simp: pred_conj_def)
-   apply fastforce
+           (fastforce dest: nf cond intro: body_inv abs_termination))
+   apply (rule_tac P="\<lambda>s'. \<exists>rv s. (s, s') \<in> srel \<and> rrel rv conc_r
+                                  \<and> P rv s \<and> (P' conc_r s' \<and> C' conc_r s') \<and> s' = new_s"
+                in no_fail_pre[rotated])
+    apply fastforce
+   apply (rule no_fail_ex_pre)
+   apply (rename_tac abs_r)
+   apply (rule no_fail_pre)
+    apply (rule_tac G="rrel abs_r conc_r" in no_fail_grab_asm)
+    apply (rule corres_u_nofail)
+     apply (fastforce dest: body_corres)
+    apply (fastforce dest: nf)
+   apply (fastforce simp: cond)
   apply (fastforce intro: wf_subset[OF whileLoop_terminates_wf[where C=C']])
   done
 
+lemmas corres_whileLoop_abs =
```

Comment:
Out of curiosity, when would you reach for the _abs form of the whileLoop rule?

Reply:
I think the `_abs` form would be better in almost all circumstances. 

I suppose the non-`_abs` form is more general (it seems possible for the abstract version to not terminate, where as the concrete version might terminate in all relevant situations). 

But the `_abs` form allows you to reason about `no_fail` and `whileLoop_terminates` on the abstract side, and that is often much more convenient, both because abstract functions are generally just nicer overall, and because the data structures are nicer (termination on the abstract side might be an Isabelle list shrinking in size, whereas on the concrete side we might need to argue about heaps of next pointers having no loops, or something).